### PR TITLE
Mibm 222 persist goodsRouteDestination

### PR DESCRIPTION
**Pull request by Leeroyc720**

### CODE REVIEW

File: app/uk/gov/hmrc/merchandiseinbaggagefrontend/controllers/GoodsRouteDestinationController.scala [hunk 1]
```diff
@@ -17,37 +17,49 @@
...
-            case (_, Import) => Redirect(routes.CannotUseServiceIrelandController.onPageLoad())
-            case (_, Export) => Redirect(routes.NoDeclarationNeededController.onPageLoad())
-          }
+          updateAndRedirect(value)
       )
   }
+  def updateAndRedirect (value : YesNo)(implicit request:DeclarationJourneyRequest[AnyContent]): Future[Result] =
+    repo.upsert(request.declarationJourney.copy(maybeGoodsRouteDestination = Some(value))).map { _ =>
+      (value, request.declarationJourney.declarationType) match {
+        case (No, _) => Redirect(routes.ExciseAndRestrictedGoodsController.onPageLoad())
```
Comment (PaulHodgson) [inline app/uk/gov/hmrc/merchandiseinbaggagefrontend/controllers/GoodsRouteDestinationController.scala:59]: It looks like the routing has been changed here? This isn't part of the acceptance criteria.

File: app/uk/gov/hmrc/merchandiseinbaggagefrontend/controllers/testonly/TestOnlyController.scala [hunk 2]
```diff
@@ -81,6 +81,7 @@ object TestOnlyController {
       sessionId = sessionId,
       declarationType = DeclarationType.Import,
       maybeExciseOrRestrictedGoods = Some(No),
+      maybeGoodsRouteDestination = Some(No),
```
Comment (PaulHodgson) [inline app/uk/gov/hmrc/merchandiseinbaggagefrontend/controllers/testonly/TestOnlyController.scala:84]: This field are isn't clear - to be fair, the url and page name in the prototype are not very clear either.

The name would need to cover both import and export usage... 

Perhaps something like 'maybeGoodsAreMovingToOrFromTheEUViaNorthernIreland' ?

Comment (PaulHodgson) [reply app/uk/gov/hmrc/merchandiseinbaggagefrontend/controllers/testonly/TestOnlyController.scala:84]: Open to better suggestions for the name!

File: app/uk/gov/hmrc/merchandiseinbaggagefrontend/model/core/Declaration.scala [hunk 3]
```diff
@@ -189,6 +190,7 @@ case class DeclarationJourney(sessionId: SessionId,
   val declarationIfRequiredAndComplete: Option[Declaration] = {
     val discardedAnswersAreCompleteAndRequireADeclaration =
       maybeGoodsDestination.isDefined &&
+        maybeGoodsRouteDestination.contains(No) &&
```
Comment (PaulHodgson) [inline app/uk/gov/hmrc/merchandiseinbaggagefrontend/model/core/Declaration.scala:193]: a clearer name would make this easier to understand...

Comment (PaulHodgson) [reply app/uk/gov/hmrc/merchandiseinbaggagefrontend/model/core/Declaration.scala:193]: However maybeGoodsRouteDestination is an optional question. It only has to be answered if  maybeGoodsDestination.contains(NorthernIreland).

Think this was a problem with the acceptance criteria.

I suspect this change makes it impossible to complete a declaration which doesn't go via Northern Ireland?

Comment (PaulHodgson) [reply app/uk/gov/hmrc/merchandiseinbaggagefrontend/model/core/Declaration.scala:193]: Maybe we should re-work the Acceptance Criteria...

File: test/uk/gov/hmrc/merchandiseinbaggagefrontend/pagespecs/GoodsRouteDestinationPageSpec.scala [hunk 4]
```diff
@@ -30,13 +30,13 @@ class GoodsRouteDestinationPageSpec extends DeclarationDataCapturePageSpec[YesNo
     behave like aPageWhichRenders(path, givenAnExportJourneyIsStarted(), exportTitle)
     behave like aPageWhichRequiresADeclarationJourney(path)
 
-    behave like aDataCapturePageWithConditionalRoutingWithoutPersistence(
+    behave like aDataCapturePageWithConditionalRouting(
       path, givenAnImportJourneyIsStarted(), Yes, CannotUseServiceIrelandPage.path)
-    behave like aDataCapturePageWithConditionalRoutingWithoutPersistence(
+    behave like aDataCapturePageWithConditionalRouting(
       path, givenAnImportJourneyIsStarted(), No, ExciseAndRestrictedGoodsPage.path, "and declarationType is Import")
-    behave like aDataCapturePageWithConditionalRoutingWithoutPersistence(
```
Comment (PaulHodgson) [inline test/uk/gov/hmrc/merchandiseinbaggagefrontend/pagespecs/GoodsRouteDestinationPageSpec.scala:37]: I don't think this test should be removed, as discussed above changing the routing from the page wasn't part of the A/C

File: test/uk/gov/hmrc/merchandiseinbaggagefrontend/pagespecs/GoodsRouteDestinationPageSpec.scala [hunk 5]
```diff
@@ -45,5 +45,5 @@ class GoodsRouteDestinationPageSpec extends DeclarationDataCapturePageSpec[YesNo
   // PH - I think this should be persisted on the DeclarationJourney
```
Comment (PaulHodgson) [inline test/uk/gov/hmrc/merchandiseinbaggagefrontend/pagespecs/GoodsRouteDestinationPageSpec.scala:45]: the comments could be removed as part of this ticket

Comment (Leeroyc720) [reply test/uk/gov/hmrc/merchandiseinbaggagefrontend/pagespecs/GoodsRouteDestinationPageSpec.scala:45]: will clean up 🧹 no problems, Thanks Paul

File: test/uk/gov/hmrc/merchandiseinbaggagefrontend/pagespecs/GoodsRouteDestinationPageSpec.scala [hunk 6]
```diff
@@ -45,5 +45,5 @@ class GoodsRouteDestinationPageSpec extends DeclarationDataCapturePageSpec[YesNo
...
   // and we should check the user has answered the question before /check-your-answers
   // otherwise the user could force browse and avoid answering the question, or submit a declaration for an invalid case
-  override def extractFormDataFrom(declarationJourney: DeclarationJourney): Option[YesNo] = Some(Yes)
+  override def extractFormDataFrom(declarationJourney: DeclarationJourney): Option[YesNo] = declarationJourney.maybeGoodsRouteDestination
 }
```
Comment (PaulHodgson) [inline test/uk/gov/hmrc/merchandiseinbaggagefrontend/pagespecs/GoodsRouteDestinationPageSpec.scala:49]: I think you should add some tests in here:

https://github.com/hmrc/merchandise-in-baggage-frontend/blob/a90857b04230f061caf5cee695b1e8fccc2d92c7/test/uk/gov/hmrc/merchandiseinbaggagefrontend/model/core/DeclarationSpec.scala#L262

to show that the declaration is incomplete if the destination is NI and the goods are moving to/from the EU

**Review by PaulHodgson — CHANGES_REQUESTED** [review]
I think... 
1. the routing for GoodsRouteDestinationController shouldn't change
2. the question is optional - we might need to make that clearer in the acceptance criteria.
3. we need a few extra tests

File: app/uk/gov/hmrc/merchandiseinbaggagefrontend/model/core/Declaration.scala [hunk 7]
```diff
@@ -189,6 +190,7 @@ case class DeclarationJourney(sessionId: SessionId,
   val declarationIfRequiredAndComplete: Option[Declaration] = {
     val discardedAnswersAreCompleteAndRequireADeclaration =
       maybeGoodsDestination.isDefined &&
+        maybeImportOrExportGoodsFromTheEUViaNorthernIreland.contains(No) &&
```
Comment (PaulHodgson) [inline app/uk/gov/hmrc/merchandiseinbaggagefrontend/model/core/Declaration.scala:193]: I think this is still wrong. maybeImportOrExportGoodsFromTheEUViaNorthernIreland only needs two be 'No' if maybeGoodsDestination == Some(NorthernIreland)

And so I think the tests you added below aren't quite right also

**Review by PaulHodgson — CHANGES_REQUESTED** [review]
maybeImportOrExportGoodsFromTheEUViaNorthernIreland only needs two be 'No' if maybeGoodsDestination == Some(NorthernIreland)

File: app/uk/gov/hmrc/merchandiseinbaggagefrontend/model/core/Declaration.scala [hunk 8]
```diff
@@ -188,12 +189,18 @@ case class DeclarationJourney(sessionId: SessionId,
...
+      }
+
     val discardedAnswersAreCompleteAndRequireADeclaration =
-      maybeGoodsDestination.isDefined &&
-        maybeImportOrExportGoodsFromTheEUViaNorthernIreland.contains(No) &&
+      ultimateSourceOrDestinationIsDefinedAndNotTheEUViaNorthernIreland &&
         maybeExciseOrRestrictedGoods.contains(No) &&
         maybeValueWeightOfGoodsExceedsThreshold.contains(No) &&
-        maybeCustomsAgent.isDefined || maybeIsACustomsAgent.contains(No)
+        (maybeCustomsAgent.isDefined || maybeIsACustomsAgent.contains(No))
```
Comment (PaulHodgson) [inline app/uk/gov/hmrc/merchandiseinbaggagefrontend/model/core/Declaration.scala:203]: our bug, now fixed

**Review by PaulHodgson — APPROVED** [review]

**Pull request merged**